include/range/v3/utility/swap.hpp

```diff
@@ -42,10 +53,26 @@ namespace ranges
         /// \cond
         namespace adl_swap_detail
         {
-            using std::swap;
+
+            // The following implementation of swap is adapted from libc++:
+            template <class T>
+            RANGES_CXX14_CONSTEXPR
+            typename std::enable_if<
+                std::is_move_constructible<T>::value &&
+                std::is_move_assignable<T>::value
+             >::type
+            swap(T& x, T& y)
+            noexcept(std::is_nothrow_move_constructible<T>::value &&
+                     std::is_nothrow_move_assignable<T>::value)
+            {
+                T t(std::move(x));
+                x = std::move(y);
+                y = std::move(t);
+            }
```

Comment:
So I just learned that yes, it is common to explicitly specialize std::swap since you cannot use ADL for hooking up a swap for e.g. `std::reference_wrapper<T>`. I'll rethink this.